### PR TITLE
v5: Underline links

### LINKED ISSUE #15304: Links (as part of regular text) not distinguishable enough
For aesthetic reasons, by default Bootstrap links have no underlines, relying instead just on a color difference from surrounding text.

This in itself violates WCAG 2.0 "1.4.1 Use of color" http://www.w3.org/TR/WCAG20/#visual-audio-contrast-without-color

There is a technique suggesting that sufficient difference in color contrast/luminance may be enough in certain conditions, but having some additional cue is preferred http://www.w3.org/TR/2014/NOTE-WCAG20-TECHS-20140916/G183

There are also situations where links are even less distinguishable.

bootlink2

http://getbootstrap.com/components/#alerts-links - link is of course distinguishable from surrounding text, but the difference between the actionable link and the bold text at the start of the alert ("Well done!" etc) is again only down to a subtle variation in color ... this is confusing for users, as it's not immediately clear what is and isn't a link.

bootlink3

http://getbootstrap.com/components/#navbar-links No distinction at all between inert text and link, unless the user goes on a "let me just hover my mouse over every word" pixel hunt. All sighted users are affected by this (though at least keyboard users will more readily discover these links by virtue of tabbing onto them).

In short: I propose that by default, links should ideally have some subtle additional indication that they're links. Of course, for links in navigation bars etc, this does not apply...the context there is clear enough ("this is a series of actionable things"). But particularly for links that are just part of regular text, we should really do something.

Failing that, this needs to be noted in our accessibility guidance.

/cc @cvrebert

### PR DESCRIPTION
**Pull request by mdo**
This PR changes the `$link-text-decoration` from `none` to `underline` to help improve link discovery. It also includes a fix to the docs sidebar to override that for our desired look. I've kind of missed underlines on our links throughout the docs at least, so this feels like a win-win to me.

With the work done in #30262, this also doesn't affect any existing Bootstrap component. All the components you'd expect will nullify the underline thanks to the changes from that PR.

Fixes #15304.

---

Let me know your thoughts @patrickhlauke and @twbs/css-review.

Preview: <https://deploy-preview-30389--twbs-bootstrap.netlify.com/>

### CODE REVIEW

File: scss/_variables.scss [hunk 1]
```diff
@@ -249,9 +249,9 @@ $body-text-align:           null !default;
 // Style anchor elements.
 
 $link-color:                              $primary !default;
-$link-decoration:                         none !default;
+$link-decoration:                         underline !default;
 $link-hover-color:                        darken($link-color, 15%) !default;
-$link-hover-decoration:                   underline !default;
+$link-hover-decoration:                   $link-decoration !default;
```
Comment (MartijnCuppens) [inline scss/_variables.scss]: We could `null`ify this. This way we could also remove the hover text-decorations from buttons.

File: site/assets/scss/_sidebar.scss [hunk 2]
```diff
@@ -24,6 +24,7 @@
   padding: .25rem .625rem .25rem .5rem;
   font-weight: 600;
   color: rgba($black, .65);
+  text-decoration: none;
```
Comment (MartijnCuppens) [inline site/assets/scss/_sidebar.scss]: By adding this, line 35 becomes redundant. Same for line 74.

**Comment by MartijnCuppens** [comment]
The hover text-decorations covered in https://github.com/twbs/bootstrap/pull/30262/files can now be `null`ified if `$link-hover-decoration == null`

**Comment by XhmikosR** [comment]
Should we remove the effect for ToC links?

Generally I can't tell if I like it or not, I guess it takes some time to evaluate this since I was used to no underline.

**Comment by MartijnCuppens** [comment]
> Should we remove the effect for ToC links?

They are going to rethemed anyway: https://github.com/twbs/bootstrap/pull/30354

> Generally I can't tell if I like it or not, I guess it takes some time to evaluate this since I was used to no underline.

You'll get used to it. It's way clearer for visually impared people.

**Comment by patrickhlauke** [comment]
from a cursory glance on the netlify preview...i like it (but yes should be rethemed/suppressed for the TOCs)

**Comment by ysds** [comment]
I agree with @XhmikosR, but anyway, we have time before the official v5 release, so let's get feedback from users.

**Comment by mdo** [comment]
Updated this the other day so it's ready to go I think.

**Comment by XhmikosR** [comment]
I think we should wait a bit on this. For example, I see `text-muted` has no visual difference when focusing or hovering. Not sure if it's intentional. Also, there might be more such cases.

**Comment by MartijnCuppens** [comment]
I've pushed some changes I mentioned in https://github.com/twbs/bootstrap/pull/30389#issuecomment-598388393.

> I see text-muted has no visual difference when focusing or hovering.

`text-muted` shouldn't have a hover state, the [colored links](http://localhost:9001/docs/4.3/helpers/colored-links/) should be used for this.

We might need to revisit the color change of hovered links later, since they are always darkened. See
https://github.com/twbs/bootstrap/blob/81223ee0412cc937cde108242e93b4407f6cede5/scss/helpers/_colored-links.scss#L8

**Review by MartijnCuppens — APPROVED** [review]

**Comment by XhmikosR** [comment]
> `text-muted` shouldn't have a hover state, the [colored links](http://localhost:9001/docs/4.3/helpers/colored-links/) should be used for this.

We need to fix this in our docs then. I find it wrong that in the homepage, where we use text-muted, there's no focus/hover style.

**Comment by patrickhlauke** [comment]
>  I find it wrong that in the homepage, where we use text-muted, there's no focus/hover style.

for focus, we at least have the default outline though, just to clarify.

**Review by patrickhlauke — APPROVED** [review]
not checked all possible variations/edge cases, but this looks good to me from looking over the deploy preview

**Comment by XhmikosR** [comment]
>for focus, we at least have the default outline though, just to clarify.

True, TBH I didn't even notice it :P

Anyway, I'm not blocking the patch :)

**Comment by MartijnCuppens** [comment]
I'll open a new PR to tackle the hover color of the link on the homepage, we can merge this if all checks succeed

**Comment by patrickhlauke** [comment]
> I'll open a new PR to tackle the hover color of the link on the homepage

be good to check that various situations (are there any more, beyond `.text-muted`?) are also covered in general

**Review by ysds — APPROVED** [review]

**Pull request merged**